File: electron-src/cayo.ts
import * as jimp from 'jimp'
import * as nut from '@nut-tree-fork/nut-js'
import * as path from 'path'
import * as fs from 'fs'
import {
    screen,
    imageSimilarity,
    press,
    wait,
    findImgInArray,
    getScreenSize
} from './utils'
import {
    UPDATE_RATE,
    CAYO_FINGERPRINT_COUNT,
    CAYO_FINGERPRINT_POS,
    CAYO_HEADER_POS,
    CAYO_PARTS_POS,
    MOVE_DELAY
} from './constants'

/**
 * load the full images of a number of fingerprints
 * @param count number of fingerprint
 * @returns an array of images
 */
async function loadFingerprints(
    count: number,
    height: number
): Promise<jimp[]> {
    return await Promise.all(
        new Array(count)
            .fill(0)
            .map((_, i) =>
                jimp.read(
                    path.join(
                        __dirname,
                        '..',
                        'assets',
                        height + '',
                        'cayo',
                        `${i + 1}`,
                        'fingerprint.png'
                    )
                )
            )
    )
}

/**
 * load the parts to check for a number of fingerprints
 * @param count number of fingerprint
 * @returns an array of array of images
 */
async function loadFingerprintParts(
    count: number,
    height: number
): Promise<jimp[][]> {
    let res: jimp[][] = []
    for (let index = 0; index < count; index++) {
        res.push(
            await Promise.all(
                new Array(8)
                    .fill(0)
                    .map((_, i) =>
                        jimp.read(
                            path.join(
                                __dirname,
                                '..',
                                'assets',
                                height + '',
                                'cayo',
                                `${index + 1}`,
                                `${i + 1}.png`
                            )
                        )
                    )
            )
        )
    }
    return res
}

;(async () => {
    const [_, height] = await getScreenSize()

    const headerIMG = await jimp.read(
        path.join(__dirname, '..', 'assets', height + '', 'cayo', 'header.png')
    )

    const FINGERPRINT_POS = CAYO_FINGERPRINT_POS[height]
    const PARTS_POS = CAYO_PARTS_POS[height]
    const HEADER_POS = CAYO_HEADER_POS[height]

    const fingerprints = await loadFingerprints(CAYO_FINGERPRINT_COUNT, height)
    const fingerprintsParts = await loadFingerprintParts(
        CAYO_FINGERPRINT_COUNT,
        height
    )

    console.log('waiting for cayo fingerprint ...')

    /*await wait(5000)

    await screenGraber()*/

    while (true) {
        const headerScreenshot = await screen(HEADER_POS)
        if (imageSimilarity(headerScreenshot, headerIMG) < 0.1) {
            const fingerprintScreenshot = await screen(FINGERPRINT_POS)

            const [_, fingerprintIndex] = findImgInArray(
                fingerprintScreenshot,
                fingerprints
            )

            console.log('fingerprint detected : ', fingerprintIndex + 1)

            const currentFingerprintParts = fingerprintsParts[fingerprintIndex]

            for (let i = 0; i < 8; i++) {
                const current_pos = PARTS_POS[i]
                const beforeScreen = new Date().getTime()
                const currentPartScreen = await screen(current_pos)
                const afterScreen = new Date().getTime()

                const [currentPartSimilarity, currentPartIndex] =
                    findImgInArray(currentPartScreen, currentFingerprintParts)

                const afterDistance = new Date().getTime()

                console.log('screen time', afterScreen - beforeScreen)
                console.log('distance time', afterDistance - afterScreen)
                console.log(
                    `part ${i + 1} : ${
                        currentPartIndex + 1
                    } with similarity ${currentPartSimilarity}`
                )

                const beforeMove = new Date().getTime()
                await moveTo(currentPartIndex, i)
                const afterMove = new Date().getTime()
                console.log('move time', afterMove - beforeMove)
                await press(nut.Key.Down)
                await wait(MOVE_DELAY)
                console.log('----')
            }
            console.log(
                '------------------------------------------------------------------------------'
            )
            await wait(2700)
        }
        await wait(1000 / UPDATE_RATE)
    }
})()

async function moveTo(current: number, target: number) {
    if (current === target) return

    if (target > current) {
        if (target - current > 4) {
            const moveCount = 8 - target + current
            console.log(`moving left ${moveCount} time(s)`)
            for (let j = 0; j < moveCount; j++) {
                await press(nut.Key.Left)
                await wait(MOVE_DELAY)
            }
        } else {
            const moveCount = target - current
            console.log(`moving right ${moveCount} time(s)`)
            for (let j = 0; j < moveCount; j++) {
                await press(nut.Key.Right)
                await wait(MOVE_DELAY)
            }
        }
    } else {
        if (target - current < -4) {
            const moveCount = 8 - current + target
            console.log(`moving right ${moveCount} time(s)`)
            for (let j = 0; j < moveCount; j++) {
                await press(nut.Key.Right)
                await wait(MOVE_DELAY)
            }
        } else {
            const moveCount = current - target
            console.log(`moving left ${moveCount} time(s)`)
            for (let j = 0; j < moveCount; j++) {
                await press(nut.Key.Left)
                await wait(MOVE_DELAY)
            }
        }
    }
}

async function screenGrabber() {
    const [_, height] = await getScreenSize()
    const HEADER_POS = CAYO_HEADER_POS[height]
    const FINGERPRINT_POS = CAYO_FINGERPRINT_POS[height]
    const PARTS_POS = CAYO_PARTS_POS[height]

    const outPath = path.join(__dirname, '..', 'out', 'cayoScreenshots')

    const headerIMG = await jimp.read(
        path.join(__dirname, '..', 'assets', height + '', 'cayo', 'header.png')
    )
    while (true) {
        const headerScreenshot = await screen(HEADER_POS)
        if (imageSimilarity(headerScreenshot, headerIMG) < 0.1) {
            const fullScreenshot = await screen([0, 0, 1920, 1080])
            const fingerprintScreenshot = await screen(FINGERPRINT_POS)

            // folder count in outPath
            const folders = fs
                .readdirSync(outPath)
                .filter(file =>
                    fs.statSync(path.join(outPath, file)).isDirectory()
                )

            let exists = false
            for (const folder of folders) {
                const folderPath = path.join(outPath, folder)
                if (!fs.existsSync(path.join(folderPath, 'fingerprint.png'))) {
                    continue
                }
                const full = await jimp.read(
                    path.join(folderPath, 'fingerprint.png')
                )
                const similarity = imageSimilarity(full, fingerprintScreenshot)
                if (similarity < 0.1) {
                    console.log('fingerprint already saved')
                    exists = true
                    break
                }
            }

            if (!exists) {
                const folderPath = path.join(outPath, `${folders.length + 1}`)
                fs.mkdirSync(folderPath)
                await fingerprintScreenshot.writeAsync(
                    path.join(folderPath, 'fingerprint.png')
                )
                await fullScreenshot.writeAsync(
                    path.join(folderPath, 'full.png')
                )

                // screenshot the parts
                for (let i = 0; i < 8; i++) {
                    const current_pos = PARTS_POS[0]
                    await press(nut.Key.Down)
                    await wait(50)
                    const currentPartScreen = await screen(current_pos)
                    await currentPartScreen.writeAsync(
                        path.join(folderPath, `${i + 1}.png`)
                    )
                    await press(nut.Key.Up)
                    await wait(50)
                    await press(nut.Key.Right)
                    await wait(50)
                }

                console.log('fingerprint saved')
            }
        }
        await wait(1000)
    }
}

File: src-tauri/src/cayo/cayo.rs
use gta_assistant::{constants, utils, ThreadStatus};
use image::RgbImage;
use std::{path::Path, thread, time::Instant};
use winapi::um::winuser::{VK_DOWN, VK_LEFT, VK_RIGHT};

pub fn handler(thread_status: ThreadStatus) {
    thread::spawn(move || {
        let height = utils::get_main_monitor().unwrap().height();
        let asset_folder = Path::new("../assets").join(height.to_string()).join("cayo");

        let header_pos = constants::CAYO_HEADER_POS.get(&height).unwrap();
        let fingerprint_pos = constants::CAYO_FINGERPRINT_POS.get(&height).unwrap();
        let parts_pos = constants::CAYO_PARTS_POS.get(&height).unwrap();

        let header_image = utils::open_image(asset_folder.join("header.png"));

        let fingerprints: Vec<RgbImage> = (1..=*constants::CAYO_FINGERPRINT_COUNT)
            .map(|i| utils::open_image(asset_folder.join(i.to_string()).join("fingerprint.png")))
            .collect();

        let parts: Vec<Vec<RgbImage>> = (1..=*constants::CAYO_FINGERPRINT_COUNT)
            .map(|fingerprint| {
                (1..=8)
                    .map(|part| {
                        utils::open_image(
                            asset_folder
                                .join(fingerprint.to_string())
                                .join(format!("{}.png", part)),
                        )
                    })
                    .collect()
            })
            .collect();

        let monitor = utils::get_main_monitor().unwrap();
        loop {
            if !utils::check_thread_status(&thread_status) {
                break;
            };

            let header_screenshot = utils::capture_region(&monitor, header_pos).into_rgb8();
            let similarity = utils::compare_image(&header_image, &header_screenshot);
            if similarity > 0.99 {
                println!("Fingerprint detected ({} header similarity)", similarity);
                let before_screens = Instant::now();
                let fingerprint_screenshot =
                    utils::capture_region(&monitor, fingerprint_pos).into_rgb8();

                let time1 = Instant::now();
                let fingerprint_index =
                    utils::find_image_in_array(&fingerprint_screenshot, &fingerprints);
                println!(
                    "fingerprint detect time : {}ms",
                    time1.elapsed().as_millis()
                );
                let curr_parts = parts.get(fingerprint_index).unwrap();
                let time2 = Instant::now();
                let parts_screenshots: Vec<RgbImage> = utils::capture_regions(&monitor, &parts_pos)
                    .into_iter()
                    .map(|img| img.into_rgb8())
                    .collect();

                println!("screen time : {}ms", before_screens.elapsed().as_millis());
                println!("only parts screen time : {}ms", time2.elapsed().as_millis());
                for i in 0..8 {
                    let part_screen = parts_screenshots.get(i).unwrap();
                    let part_index = utils::find_image_in_array(part_screen, &curr_parts);
                    println!(
                        "part n°{} : current index: {}, target index : {}",
                        i, part_index, i
                    );

                    move_to(part_index, i);
                    utils::press(VK_DOWN);
                }
                thread::sleep(*constants::CAYO_WAIT_DELAY - *constants::LOOP_DELAY);
            }
            thread::sleep(*constants::LOOP_DELAY);
        }
    });
}

fn move_to(current: usize, target: usize) {
    if current == target {
        return;
    }

    if target > current {
        if target - current > 4 {
            let move_count = 8 - target + current;
            multiple_press(VK_LEFT, move_count);
        } else {
            let move_count = target - current;
            multiple_press(VK_RIGHT, move_count);
        }
    } else {
        if current - target > 4 {
            let move_count = 8 - current + target;
            multiple_press(VK_RIGHT, move_count);
        } else {
            let move_count = current - target;
            multiple_press(VK_LEFT, move_count);
        }
    }
}

fn multiple_press(key: i32, count: usize) {
    for _ in 0..count {
        utils::press(key);
        // MAYBE ADD DELAY
    }
}
